src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs

```diff
@@ -698,48 +685,50 @@ private HubMessage BindInvocationMessage(string invocationId, string target, obj
             return new InvocationMessage(invocationId, target, arguments, streamIds);
         }
 
-        private object BindType(JsonElement jsonObject, Type type)
+        private object BindType(ref Utf8JsonReader reader, Type type)
         {
-            if (type == typeof(DateTime))
-            {
-                return jsonObject.GetDateTime();
-            }
-            else if (type == typeof(DateTimeOffset))
-            {
-                return jsonObject.GetDateTimeOffset();
-            }
-
-            return JsonSerializer.Parse(jsonObject.GetRawText(), type, _payloadSerializerOptions);
+            return JsonSerializer.ReadValue(ref reader, type, _payloadSerializerOptions);
         }
 
-        private object[] BindTypes(JsonElement jsonArray, IReadOnlyList<Type> paramTypes)
+        private object[] BindTypes(ref Utf8JsonReader reader, IReadOnlyList<Type> paramTypes)
         {
             object[] arguments = null;
             var paramIndex = 0;
-            var argumentsCount = jsonArray.GetArrayLength();
             var paramCount = paramTypes.Count;
 
-            if (argumentsCount != paramCount)
-            {
-                throw new InvalidDataException($"Invocation provides {argumentsCount} argument(s) but target expects {paramCount}.");
-            }
+            var depth = reader.CurrentDepth;
+            reader.CheckRead();
 
-            foreach (var element in jsonArray.EnumerateArray())
+            while (reader.TokenType != JsonTokenType.EndArray && reader.CurrentDepth > depth)
             {
                 if (arguments == null)
                 {
                     arguments = new object[paramCount];
                 }
 
-                try
+                if (paramIndex < paramCount)
```

Comment:
L702 can be simplified and brought inside the look for more pranav points. Then you avoid the allocation with a zero-parameter method in the case that there's data in the array.